File: zig/src/token.zig
const std = @import("std");

pub const TokenType = enum {
    ILLEGAL,
    EOF,

    // Identifiers + Literals
    IDENT,
    INT,
    STRING,

    // Operators
    ASSIGN,
    PLUS,
    MINUS,
    BANG,
    ASTERISK,
    SLASH,

    LT,
    GT,

    EQ,
    NOT_EQ,

    // Delimiters
    COMMA,
    SEMICOLON,
    LPAREN,
    RPAREN,
    LBRACE,
    RBRACE,

    // Keywords
    FUNCTION,
    LET,
    TRUE,
    FALSE,
    IF,
    ELSE,
    RETURN,
};

pub const Token = struct {
    type: TokenType,
    literal: []const u8,

    pub fn lookupIdent(ident: []const u8) TokenType {
        if (std.mem.eql(u8, ident, "fn")) return .FUNCTION;
        if (std.mem.eql(u8, ident, "let")) return .LET;
        if (std.mem.eql(u8, ident, "true")) return .TRUE;
        if (std.mem.eql(u8, ident, "false")) return .FALSE;
        if (std.mem.eql(u8, ident, "if")) return .IF;
        if (std.mem.eql(u8, ident, "else")) return .ELSE;
        if (std.mem.eql(u8, ident, "return")) return .RETURN;
        return .IDENT;
    }
};

File: rust/src/token.rs
#[derive(Debug, PartialEq, Clone)]
pub enum Token {
    // Keywords
    Let,
    Fn,
    If,
    Else,
    Return,
    True,
    False,

    // Identifiers + Literals
    Ident(String),
    Int(i64),
    String(String),

    // Operators
    Assign,
    Plus,
    Minus,
    Bang,
    Asterisk,
    Slash,
    Equal,
    NotEqual,
    LessThan,
    GreaterThan,
    
    // Delimiters
    Comma,
    Semicolon,
    LParen,
    RParen,
    LBrace,
    RBrace,

    // End of File
    Illegal,
    EOF,
}

impl Token {
    pub fn lookup_ident(ident: &str) -> Token {
        match ident {
            "fn" => Token::Fn,
            "let" => Token::Let,
            "true" => Token::True,
            "false" => Token::False,
            "if" => Token::If,
            "else" => Token::Else,
            "return" => Token::Return,
            _ => Token::Ident(ident.to_string()),
        }
    }
}
